File: online/cloud_cache.py
"""Sessão guardada, cache local dos saves da conta e períodos da leaderboard."""

import json
import os
import re
import time
import uuid

from config import SAVE_DIR
from i18n import tr
from online.firebase import (
    LEADERBOARD_FETCH_DELAY, LEADERBOARD_PERIOD, LEADERBOARD_SIZE, save_summary,
)


SESSION_PATH = os.path.join(SAVE_DIR, "lucky_verities_session.json")
CACHE_DIR = os.path.join(SAVE_DIR, "cloud_cache")
LEADERBOARD_CACHE_PATH = os.path.join(SAVE_DIR, "lucky_verities_leaderboard.json")


# ---- sessão (para não teres de iniciar sessão de cada vez que abres o jogo) ----
def load_session():
    try:
        with open(SESSION_PATH, "r", encoding="utf-8") as f:
            d = json.load(f)
        if all(isinstance(d.get(k), str) and d.get(k) for k in ("uid", "username", "refresh_token")):
            return d
    except (OSError, ValueError, AttributeError):
        pass
    return None


def store_session(uid, username, refresh_token, last_pub_time=0.0):
    try:
        tmp = SESSION_PATH + ".tmp"
        with open(tmp, "w", encoding="utf-8") as f:
            json.dump({"uid": uid, "username": username, "refresh_token": refresh_token,
                       "last_pub_time": float(last_pub_time)}, f)
        os.replace(tmp, SESSION_PATH)
    except OSError:
        pass


def clear_session():
    try:
        if os.path.exists(SESSION_PATH):
            os.remove(SESSION_PATH)
    except OSError:
        pass


# ---- nome deste PROCESSO (para o "uma conta, um jogo de cada vez") ----
INSTALL_ID_RE = re.compile(r"^[A-Za-z0-9_-]{8,64}$")


def install_id():
    """Identificador NOVO a cada arranque do jogo (não fica guardado em disco). Assim, dois .exe
    abertos ao mesmo tempo - mesmo no mesmo PC - contam sempre como dispositivos diferentes, e um
    bloqueia mesmo o outro. Fechar o jogo normalmente larga logo a marca (ver release_session), por
    isso reabrir a seguir no mesmo PC continua instantâneo - só um crash é que obriga a esperar os
    SESSION_STALE segundos, tal como acontecia antes."""
    return uuid.uuid4().hex


# ---- cache local dos saves da conta (o jogo grava aqui de 10 em 10 s e sincroniza com a cloud) ----
def cache_path(uid, slot, suffix=""):
    safe = re.sub(r"[^A-Za-z0-9_-]", "", str(uid)) or "unknown"
    return os.path.join(CACHE_DIR, safe, "slot%d%s.json" % (slot, suffix))


def read_cache(uid, slot):
    try:
        with open(cache_path(uid, slot), "r", encoding="utf-8") as f:
            d = json.load(f)
        if isinstance(d, dict) and isinstance(d.get("state"), dict):
            return {"state": d["state"], "base_time": d.get("base_time"), "dirty": bool(d.get("dirty", True))}
    except (OSError, ValueError, TypeError):
        pass
    return None


def write_cache(uid, slot, base_time, dirty, state_dict, suffix=""):
    path = cache_path(uid, slot, suffix)
    try:
        os.makedirs(os.path.dirname(path), exist_ok=True)
        tmp = path + ".tmp"
        with open(tmp, "w", encoding="utf-8") as f:
            json.dump({"base_time": base_time, "dirty": bool(dirty), "state": state_dict}, f)
        os.replace(tmp, path)
    except OSError:
        pass


def delete_cache(uid, slot):
    try:
        path = cache_path(uid, slot)
        if os.path.exists(path):
            os.remove(path)
    except OSError:
        pass


def peek_cache(uid, slot):
    c = read_cache(uid, slot)
    if not c:
        return None
    s = save_summary(c["state"])
    return {"coins": s["coins"], "total_rolls": s["rolls"], "playtime": s["playtime"], "rebirths": s["rebirths"]}


def backup_state(uid, slot, label, state_dict):
    """Guarda a versão que NÃO foi usada num conflito (nunca é apagada)."""
    write_cache(uid, slot, None, False, state_dict, suffix="_%s_backup_%d" % (label, int(time.time())))


def pick_save(cloud, cache):
    """Decide de onde vem o progresso ao abrir um slot da conta.
    cloud: {"data": dict, "time": str} ou None      cache: {"state", "base_time", "dirty"} ou None
    Devolve (state_dict|None, base_time, dirty, nota, perdedor). 'perdedor' = (nome, dict) da versão que
    NÃO foi usada quando as duas mudaram: fica num ficheiro de backup, nunca se perde."""
    if cloud is None and cache is None:
        return None, None, True, None, None
    if cache is None:
        return cloud["data"], cloud["time"], False, None, None
    if cloud is None:
        return cache["state"], None, True, None, None
    if cache["base_time"] == cloud["time"]:
        if cache["dirty"]:
            return cache["state"], cloud["time"], True, None, None
        return cloud["data"], cloud["time"], False, None, None
    if not cache["dirty"]:
        return cloud["data"], cloud["time"], False, None, None
    # a cloud mudou noutro PC E aqui também há progresso por enviar: ganha quem tem mais playtime
    try:
        local_pt = float(cache["state"].get("playtime", 0.0))
        cloud_pt = float(cloud["data"].get("playtime", 0.0))
    except (TypeError, ValueError, AttributeError):
        local_pt = cloud_pt = 0.0
    if local_pt > cloud_pt:
        return (cache["state"], cloud["time"], True,
                tr("Kept this PC's progress (it was ahead of the cloud). The cloud version was backed up."),
                ("cloud", cloud["data"]))
    return (cloud["data"], cloud["time"], False,
            tr("The cloud save was ahead of this PC. Your local version was backed up."),
            ("local", cache["state"]))


# ---- leaderboard: períodos de 10 min certinhos no relógio ----
def lb_publish_period(now):
    return int(now // LEADERBOARD_PERIOD)


def lb_fetch_period(now):
    return int((now - LEADERBOARD_FETCH_DELAY) // LEADERBOARD_PERIOD)


def lb_next_update(now):
    return (lb_fetch_period(now) + 1) * LEADERBOARD_PERIOD + LEADERBOARD_FETCH_DELAY


# tabelas da leaderboard (chave usada na cache / no ecrã) -> campo do documento /leaderboard/{uid}
LB_FIELDS = (("money", "coins"), ("playtime", "playtime"), ("rolls", "rolls"), ("rebirths", "rebirths"))
LB_KEYS = tuple(k for k, _f in LB_FIELDS)


def fetch_leaderboards(client, period):
    data = {}
    for key, field in LB_FIELDS:
        # rebirths: só entra quem já fez pelo menos 1 (senão a tabela enchia-se de zeros)
        data[key] = client.top_entries(field, LEADERBOARD_SIZE, min_value=1 if key == "rebirths" else None)
    data["period"] = period
    data["fetched_at"] = time.time()
    return data


def lb_data_complete(d):
    """False se a cache for de uma versão antiga (sem as tabelas Rolls / Rebirths) -> pede-se de novo."""
    return isinstance(d, dict) and all(isinstance(d.get(k), list) for k in LB_KEYS)


def load_lb_cache():
    try:
        with open(LEADERBOARD_CACHE_PATH, "r", encoding="utf-8") as f:
            d = json.load(f)
        if isinstance(d, dict) and isinstance(d.get("money"), list) and isinstance(d.get("playtime"), list):
            return d
    except (OSError, ValueError):
        pass
    return None


def store_lb_cache(data):
    try:
        tmp = LEADERBOARD_CACHE_PATH + ".tmp"
        with open(tmp, "w", encoding="utf-8") as f:
            json.dump(data, f)
        os.replace(tmp, LEADERBOARD_CACHE_PATH)
    except OSError:
        pass

File: src/online/cloud_cache.rs
//! Saved session, local cache of the account saves, and leaderboard periods (online/cloud_cache.py).

use crate::config::save_dir;
use crate::pyjson;
use crate::storage::value_truthy;
use serde_json::{Map, Value, json};
use std::path::PathBuf;

pub fn session_path() -> PathBuf {
    save_dir().join("lucky_verities_session.json")
}
pub fn cache_dir() -> PathBuf {
    save_dir().join("cloud_cache")
}
pub fn leaderboard_cache_path() -> PathBuf {
    save_dir().join("lucky_verities_leaderboard.json")
}

fn write_atomic(path: &PathBuf, v: &Value) {
    let mut tmp = path.clone().into_os_string();
    tmp.push(".tmp");
    let tmp = PathBuf::from(tmp);
    if std::fs::write(&tmp, pyjson::dumps(v)).is_ok() {
        let _ = std::fs::rename(&tmp, path);
    }
}

#[derive(Clone, Debug)]
pub struct Session {
    pub uid: String,
    pub username: String,
    pub refresh_token: String,
    pub raw: Map<String, Value>,
}

pub fn load_session() -> Option<Session> {
    let text = std::fs::read_to_string(session_path()).ok()?;
    let Value::Object(d) = serde_json::from_str::<Value>(&text).ok()? else {
        return None;
    };
    let get = |k: &str| match d.get(k) {
        Some(Value::String(s)) if !s.is_empty() => Some(s.clone()),
        _ => None,
    };
    Some(Session { uid: get("uid")?, username: get("username")?, refresh_token: get("refresh_token")?, raw: d.clone() })
}

pub fn store_session(uid: &str, username: &str, refresh_token: &str, last_pub_time: f64) {
    let mut m = Map::new();
    m.insert("uid".into(), json!(uid));
    m.insert("username".into(), json!(username));
    m.insert("refresh_token".into(), json!(refresh_token));
    m.insert("last_pub_time".into(), pyjson::float(last_pub_time));
    write_atomic(&session_path(), &Value::Object(m));
}

pub fn clear_session() {
    let p = session_path();
    if p.exists() {
        let _ = std::fs::remove_file(p);
    }
}

/// A fresh identifier per game launch (uuid4().hex).
pub fn install_id() -> String {
    let mut b = [0u8; 16];
    if let Ok(mut f) = std::fs::File::open("/dev/urandom") {
        use std::io::Read;
        let _ = f.read_exact(&mut b);
    } else {
        let t = std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).map(|d| d.as_nanos()).unwrap_or(0);
        b = t.to_le_bytes();
    }
    b[6] = (b[6] & 0x0f) | 0x40;
    b[8] = (b[8] & 0x3f) | 0x80;
    b.iter().map(|x| format!("{:02x}", x)).collect()
}

pub fn cache_path(uid: &str, slot: i64, suffix: &str) -> PathBuf {
    let mut safe: String = uid.chars().filter(|c| c.is_ascii_alphanumeric() || *c == '_' || *c == '-').collect();
    if safe.is_empty() {
        safe = "unknown".into();
    }
    cache_dir().join(safe).join(format!("slot{}{}.json", slot, suffix))
}

#[derive(Clone, Debug)]
pub struct CacheEntry {
    pub state: Value,
    pub base_time: Option<String>,
    pub base_time_raw: Value,
    pub dirty: bool,
}

pub fn read_cache(uid: &str, slot: i64) -> Option<CacheEntry> {
    let text = std::fs::read_to_string(cache_path(uid, slot, "")).ok()?;
    let Value::Object(d) = serde_json::from_str::<Value>(&text).ok()? else {
        return None;
    };
    let state = d.get("state")?;
    if !state.is_object() {
        return None;
    }
    let raw = d.get("base_time").cloned().unwrap_or(Value::Null);
    Some(CacheEntry {
        state: state.clone(),
        base_time: raw.as_str().map(|s| s.to_string()),
        base_time_raw: raw,
        dirty: d.get("dirty").map(value_truthy).unwrap_or(true),
    })
}

pub fn write_cache_suffix(uid: &str, slot: i64, base_time: Option<&str>, dirty: bool, state: &Value, suffix: &str) {
    let path = cache_path(uid, slot, suffix);
    if let Some(dir) = path.parent() {
        if std::fs::create_dir_all(dir).is_err() {
            return;
        }
    }
    let mut m = Map::new();
    m.insert("base_time".into(), base_time.map(|s| json!(s)).unwrap_or(Value::Null));
    m.insert("dirty".into(), json!(dirty));
    m.insert("state".into(), state.clone());
    write_atomic(&path, &Value::Object(m));
}

pub fn write_cache(uid: &str, slot: i64, base_time: Option<&str>, dirty: bool, state: &Value) {
    write_cache_suffix(uid, slot, base_time, dirty, state, "");
}

pub fn delete_cache(uid: &str, slot: i64) {
    let p = cache_path(uid, slot, "");
    if p.exists() {
        let _ = std::fs::remove_file(p);
    }
}

pub fn backup_state(uid: &str, slot: i64, label: &str, state: &Value) {
    let now = crate::core::state::now_ts() as i64;
    write_cache_suffix(uid, slot, None, false, state, &format!("_{}_backup_{}", label, now));
}

pub struct CloudSave {
    pub data: Value,
    pub time: Option<String>,
}

pub struct Picked {
    pub state: Option<Value>,
    pub base_time: Option<String>,
    pub dirty: bool,
    pub note: Option<String>,
    pub loser: Option<(&'static str, Value)>,
}

pub fn pick_save(cloud: Option<&CloudSave>, cache: Option<&CacheEntry>) -> Picked {
    let p = |state: Option<Value>, bt: Option<String>, dirty: bool| Picked { state, base_time: bt, dirty, note: None, loser: None };
    match (cloud, cache) {
        (None, None) => p(None, None, true),
        (Some(c), None) => p(Some(c.data.clone()), c.time.clone(), false),
        (None, Some(k)) => p(Some(k.state.clone()), None, true),
        (Some(c), Some(k)) => {
            let same = match (&k.base_time_raw, &c.time) {
                (Value::String(a), Some(b)) => a == b,
                (Value::Null, None) => true,
                _ => false,
            };
            if same {
                if k.dirty {
                    return p(Some(k.state.clone()), c.time.clone(), true);
                }
                return p(Some(c.data.clone()), c.time.clone(), false);
            }
            if !k.dirty {
                return p(Some(c.data.clone()), c.time.clone(), false);
            }
            let pt = |v: &Value| -> Option<f64> {
                match v.as_object()?.get("playtime") {
                    None => Some(0.0),
                    Some(x) => crate::storage::value_f64(x),
                }
            };
            let (local_pt, cloud_pt) = match (pt(&k.state), pt(&c.data)) {
                (Some(a), Some(b)) => (a, b),
                _ => (0.0, 0.0),
            };
            if local_pt > cloud_pt {
                return Picked {
                    state: Some(k.state.clone()),
                    base_time: c.time.clone(),
                    dirty: true,
                    note: Some(crate::i18n::tr("Kept this PC's progress (it was ahead of the cloud). The cloud version was backed up.")),
                    loser: Some(("cloud", c.data.clone())),
                };
            }
            Picked {
                state: Some(c.data.clone()),
                base_time: c.time.clone(),
                dirty: false,
                note: Some(crate::i18n::tr("The cloud save was ahead of this PC. Your local version was backed up.")),
                loser: Some(("local", k.state.clone())),
            }
        }
    }
}

#[derive(Clone, Debug)]
pub struct CachePeek {
    pub coins: f64,
    pub total_rolls: i64,
    pub playtime: f64,
    pub rebirths: i64,
}

pub fn peek_cache(uid: &str, slot: i64) -> Option<CachePeek> {
    let c = read_cache(uid, slot)?;
    let s = crate::online::firebase::save_summary(&c.state);
    Some(CachePeek { coins: s.coins, total_rolls: s.rolls, playtime: s.playtime, rebirths: s.rebirths })
}

// ---- leaderboard: 10-minute periods on the clock ----
use crate::online::firebase::{Client, LEADERBOARD_FETCH_DELAY, LEADERBOARD_PERIOD, LEADERBOARD_SIZE, Res};

pub fn lb_publish_period(now: f64) -> i64 {
    (now / LEADERBOARD_PERIOD).floor() as i64
}
pub fn lb_fetch_period(now: f64) -> i64 {
    ((now - LEADERBOARD_FETCH_DELAY) / LEADERBOARD_PERIOD).floor() as i64
}
pub fn lb_next_update(now: f64) -> f64 {
    (lb_fetch_period(now) + 1) as f64 * LEADERBOARD_PERIOD + LEADERBOARD_FETCH_DELAY
}

pub const LB_FIELDS: [(&str, &str); 4] = [("money", "coins"), ("playtime", "playtime"), ("rolls", "rolls"), ("rebirths", "rebirths")];

pub fn fetch_leaderboards(client: &Client, period: i64) -> Res<Value> {
    let mut data = Map::new();
    for (key, field) in LB_FIELDS {
        let entries = client.top_entries(field, LEADERBOARD_SIZE, if key == "rebirths" { Some(1) } else { None })?;
        let arr: Vec<Value> = entries
            .into_iter()
            .map(|e| {
                let mut m = Map::new();
                m.insert("username".into(), json!(e.username));
                m.insert("value".into(), pyjson::float(e.value));
                Value::Object(m)
            })
            .collect();
        data.insert(key.into(), Value::Array(arr));
    }
    data.insert("period".into(), json!(period));
    data.insert("fetched_at".into(), pyjson::float(crate::core::state::now_ts()));
    Ok(Value::Object(data))
}

pub fn lb_data_complete(d: &Value) -> bool {
    d.is_object() && LB_FIELDS.iter().all(|(k, _)| d.get(k).is_some_and(|v| v.is_array()))
}

pub fn load_lb_cache() -> Option<Value> {
    let t = std::fs::read_to_string(leaderboard_cache_path()).ok()?;
    let d: Value = serde_json::from_str(&t).ok()?;
    if d.is_object() && d.get("money").is_some_and(|v| v.is_array()) && d.get("playtime").is_some_and(|v| v.is_array()) {
        return Some(d);
    }
    None
}

pub fn store_lb_cache(data: &Value) {
    write_atomic(&leaderboard_cache_path(), data);
}
